Turn averaging into a composite material model and add normalized weighted distance averages

MFraters opened the pull request.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,148 @@
+/*
+  Copyright (C) 2014 by the authors of the ASPECT code.
```
The date is wrong. Please also just use the same formatting as we do in the other files. Just copy the whole block from somewhere else.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,148 @@
+
+
+    /**
+     * A material model that applies a depth-dependent viscosity to a ''base model''
```
This comment is out of date.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,148 @@
+      public:
+        /**
+                 * Given the averaging @p operation, a description of where the
+                 * quadrature points are located on the given cell, and a mapping,
```
please indent the comment.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,148 @@
+        void
+        average (const std::string &averaging_operation,
+                 const std::vector<Point<dim> >    &position,
+                 std::vector<double>           &values_out) const;
```
Does this function have to virtual?

It is also essentially an internal function, so you should be able to put it into the private section of the class.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,148 @@
+                 * Bellshape limit
+                 */
+        double bell_shape_limit;
+        std::string averaging_operation;
```
same here -- these should be private.

bangerth replied:
also, I think you should introduce an enum that stores the different averaging operations. The `averaging_operation` variable would then be of this enum type, and you set it to its correct value based on the string variable you get in `parse_parameters()`.

This has the advantage that you don't need to do string comparisons in `evaluate()` -- only integer comparisons with the enum necessary there, which you could even do with a `switch-case` statement.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+      std::string averaging_operation_lower = averaging_operation;
+
+      // switch to lower case, so the input becomes independent of case
+      std::transform(averaging_operation_lower.begin(), averaging_operation_lower.end(), averaging_operation_lower.begin(), ::tolower);
```
I don't think you need this -- just require that the user provides the string in all lower-case. In fact, anything that is not the correct spelling will not pass the `Patterns::Selection` you use in defining the parameter in `declare_parameters()`.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+        }
+      else if (averaging_operation_lower == "arithmetic average")
+        {
+          double sum = 0;
```
I would suggest to drop the averaging operations that are already implemented separately. This only duplicates code, but does not really add functionality that wasn't already there.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
```
You declared these two variables just above the loop. Move these declarations into the loop body itself, you don't need the variables outside anyway.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+            {
+              sum_value = 0;
+              sum_weights = 0;
+              double h = bell_shape_limit;
```
Unless you plan to update `h`, just use the original variable to avoid confusion.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+              else
+                {
+                  alfad = 106/(16 * numbers::PI * h * h *h);
+                }
```
you can write these many lines more concisely as
```
  const double alfad = (dim == 2 ? 5/(...) : 106/(...));
```
Can you add a comment explaining what this variable does (which would also help me understand the name).

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
```
same comments apply in this block as in the previous one

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+              double alfad = 0;
+              if (dim==2)
+                {
+                  alfad = 5/(numbers::PI * h * h);
```
since you have this same code multiple times, just move the declaration of `alfad` up to the function scope so you don't have to repeat the same code again.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+              double alfad = 0;
+              if (dim==2)
+                {
+                  alfad = 5/(numbers::PI * h * h);
```
since you have this variable in every block and it has the same value every time, just move it up and initialize it only once.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+                             typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      base_model -> evaluate(in,out);
+      if (out.viscosities.size() > 1)
```
you look at the viscosities here as a proxy for the sizes of the other arrays, right? If so, maybe say so in a comment.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+                            "that for more information.");
+          prm.declare_entry ("Averaging operation", "none",
+                             Patterns::Selection ("none|arithmetic average|harmonic average|geometric average|pick largest|NWD arithmetic average|NWD harmonic average|NWD geometric average"),
+                             "Chose the averaging operation to use.");
```
chose -> choose

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+                             "Chose the averaging operation to use.");
+          prm.declare_entry ("Bell shape limit", "1",
+                             Patterns::Double(0),
+                             "The limit normalized distance between 0 and 1 where the bell shape becomes zero. See the manual for a more information.");
```
-> for more information

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+        {
+          Assert( prm.get("Base model") != "averaging",
+                  ExcMessage("You may not use ``depth dependent'' as the base model for "
+                             "a depth-dependent model.") );
```
adjust error text

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+      }
+      prm.leave_subsection();
+      /* After parsing the parameters for depth dependent, it is essential to parse
+      parameters related to the base model. */
```
adjust comment.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+                                   "point which determines the weight of each quadrature point for the quadrature "
+                                   "point in question. Therefore the weight is determined by the distance between the "
+                                   "quadrature points. This distance is normalized to one, so the largest distance "
+                                   "becomes one."
```
I think a formula that explains the averaging operation would be very helpful here. I don't think I can follow the text description.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,412 @@
+                                   "point in question. Therefore the weight is determined by the distance between the "
+                                   "quadrature points. This distance is normalized to one, so the largest distance "
+                                   "becomes one."
+                                   "When NWD average models are chosen one should also set the bell shape limit."
```
Same here -- I don't know what "bell shape limit" refers to.

MFraters commented:
sorry, every time I commit I see more comments. Should learn to reload the page before committing...

bangerth commented:
No worries -- I'll show you later how to combine all of these patches again.

bangerth commented:
Let me know once you're done and I can take another look. Ideally, you can already go ahead and squash or fixup all of the commits into a single one.

bangerth commented:
Would you also mind (i) adding an entry to doc/modules/changes.h, (ii) adding a test, e.g., a modification of one of the current averaging tests.

MFraters commented:
Hey Wolfgang,
I added an entry to changes.h and added tests. For the tests I put the .prm files in the tests folder, and made a folder for each tests which contains the screen output. Is this enought, or are extra changes required to get it working?

bangerth commented on doc/modules/changes.h:
```diff
@@ -53,5 +53,10 @@
+ * <li> New: Three new material averaging schemes are added. These are combined
+ * with the existing averaging schemes, except for the Q1 averaging schemes into
+ * a compositing material model. The new averaging schemes are normalised weighted
+ * distance versions of the arithmetic, harmonic and geometric averages.
```
Add your name here.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,449 @@
+      // perform the requested averaging
+      if (averaging_operation == "none")
+        { }
+      else if (averaging_operation == "arithmetic average")
```
you still do the string comparisons. please use the enum idea I mentioned previously -- string comparisons are so expensive that we don't want to do it on every cell if it can be avoided.

MFraters replied:
Sorry, missed that comment. Fixed now.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,449 @@
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  if (position[i].distance(position[j]) > max_distance)
+                    max_distance = position[i].distance(position[j]);
```
You can write this more compactly as
```
  max_distance = std::max (max_distance,
                                            position[i].distance(position[j]));
```
This way you also don't have to compute the distance twice.

bangerth replied:
This too is still open.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,449 @@
+              for (unsigned int j=0; j<N; ++j)
+                {
+
+                  double distance = position[i].distance(position[j])/max_distance;
```
here and above, if you have a variable that is assigned to only once, mark it as `const`.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,449 @@
+                                   "of their name work with a weighed average, which means each quadrature point "
+                                   "requires an individual weight. The weight is determined by the distance, where "
+                                   "the exact relation is determined by a bell shaped curve. A bell shaped curve is "
+                                   "a continues function which is one at it's maximum and exactly zero at and beyond "
```
continues -> continuous

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,449 @@
+                                   "determine the weighting map for each quadrature point. The used bell shape comes "
+                                   "from Lucy (1977). The distance is normalized so the largest distance becomes one. "
+                                   "This means that if variable ''Bell shape limit'' is exactly one, the farthest "
+                                   "quadrature point is just on the limit and it's weight will be exactly one."
```
did you mean to say that its weight is zero?

bangerth commented:
About the tests: Yes, just adding the screen-output file is ok.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,200 @@
+    using namespace dealii;
+    //namespace AveragingOperations
+    //		{
+    /**
```
Please remove the commented out lines and indent the lines below as usual.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,200 @@
+                         * - Project to $Q_1$: This operation takes the values at the quadrature
+                         * points and computes the best bi- or trilinear approximation for them.
+                         * In other words, it projects the values into the $Q_1$ finite element
+                         * space. It then re-evaluate this projection at the quadrature points.
```
The new NWS options are not documented here.

It's ok if you remove most of the text above and simply state that the first 5 values of the enum are as in MaterialModel::AveragingOperations and that the remainder are as explained in the documentation of the `averaging model' plugin.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,200 @@
+               * AveragingOperation value.
+               */
+              AveragingOperation
+			  parse_averaging_operation_name (const std::string &s);
```
Indent as appropriate.

bangerth commented on include/aspect/material_model/averaging.h:
```diff
@@ -0,0 +1,200 @@
+        * perform this operation on all elements of the @p values structure.
+        */
+        void
+        average (const AveragingOperation &averaging_operation,
```
Passing the operation by reference does not harm, but also doesn't help very much for integer types. Leave it const but remove the reference.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,496 @@
+  std::string get_averaging_operation_names ()
+        {
+          return "none|arithmetic average|harmonic average|geometric average|pick largest|nwd arithmetic average|nwd harmonic average|nwd geometric average";
+        }*/
```
Remove if you don't need it.

bangerth commented on source/material_model/averaging.cc:
```diff
@@ -0,0 +1,496 @@
+                                   "This means that if variable ''Bell shape limit'' is exactly one, the farthest "
+                                   "quadrature point is just on the limit and it's weight will be exactly zero. In "
+								   "this plugin it is not implemented as larger and equal than the limit, but larger "
+								   "than, to ensure the the quadrature point at distance zero is always included."
```
indent.

bangerth commented:
OK. Please squash into a single commit, then it's ready to go if the tester is happy.

/run-tests
